### ISSUE TITLE
sccz80 - z80 fastmath l_long_mult

### ISSUE THREAD

**Issue by feilipu** [z88dk/z88dk#1890]
Added some additional tests https://github.com/z88dk/z88dk/commit/1a4d1358d551e462db407b64a8975e59001f764f, and found that with fast math enabled, there's a failure.

```sh
zcc +rc2014 -vn  -I../../framework -DNO_LOG_RUNNING -DNO_LOG_PASSED ../../framework/test.c mult.c -o test_rc2014_mult.bin  -clib=new -subtype=basic -m
 15:32:44  Starting rc2014/z80 test 
Starting suite Multiplication Tests (3 tests)
Running test test_mult_long.....failed mult.c:45 (-3 * -5)
3 run, 2 passed, 1 failed
 15:32:44  Test failure 
```
Works properly with small math.

```sh
zcc +rc2014 -vn  -I../../framework -DNO_LOG_RUNNING -DNO_LOG_PASSED ../../framework/test.c mult.c -o test_rc2014_mult.bin  -clib=new -subtype=basic -m
 15:35:46  Starting rc2014/z80 test 
Starting suite Multiplication Tests (3 tests)
3 run, 3 passed, 0 failed
 15:35:46  rc2014/z80 test completed 
```

**Comment by feilipu** [comment]
I can't reproduce the problem outside the test environment. i.e. the multiply returns the correct value when called from any program. Yet, the problem persists as described with the `( -3 * -5)` result causing a test failure.

Inspection of the code gives me no further leads to test. It all looks and (outside the test) behaves correctly.

So unless there is another issue with the comparison function (unlikely), I've no further ideas.

**Comment by suborb** [comment]
Hmmm. I thought I’d reproduced in a simple program. 

Let me try again later.

**Comment by suborb** [comment]
I've achieved the following:

```
#include <stdio.h>

int main() {
        long a,b,c;

        a = -5;
        b = -3;
        c = a * b;

        printf("%ld\n",c);
}
```

```
% zcc +test mult.c -debug -lfastmath
% z88dk-ticks -x a.map -debug a.bin

__head:
                    jp      program                         ;[0000] c3 3b 00

    $0000    (__head+0)> break printf
Adding breakpoint at 'printf' $0400 (printf)

    $0000    (__head+0)> cont
Hit breakpoint 1: @0400 (printf)
printf:
                    ld      l,a                             ;[0400] 6f

    $0400    (printf+0)> up
frame 1
 * function main+82 (c=50199070, b=-3, a=-5)
       at mult.c:11
```

**Comment by feilipu** [comment]
I've unfounded suspicions that the fastmath routines are not properly extending twos complement, somehow, when using the classic library. But, it is working in newlib.

The issue is not compiler related.
Add `volatile long a,b,c;` to `main.c` and the error is still there with sdcc, too.
```
$ zcc +test -compiler=sdcc main.c -lfastmath --list
$ z88dk-ticks -x a.map a.bin
Decoded cdb: <F:G$main$0_0$0({0}DF,SI:S),C,0,0,0,0,0>
Decoded cdb: <S:Lmain.c.main$a$1_0$2({4}SL:S),B,1,-4>
Decoded cdb: <S:Lmain.c.main$b$1_0$2({4}SL:S),B,1,-8>
Decoded cdb: <S:Lmain.c.main$c$1_0$2({4}SL:S),B,1,-12>
83621406

Ticks: 33091
```

**Comment by feilipu** [comment]
Ok. Tracked down the failure to the most significant byte handling.

This code tests the issue.
Above a certain 32_16x16 the result should overflow to 0.
But it doesn't.

```C
#include <stdio.h>

int main()
{
        volatile uint32_t a,b,c;
        uint8_t i;
        
        a = 1;
        b = 1;
        
        for (i = 0; i < 31; ++i)
        {
            a = a<<1;
            b = b<<1;

            c = a * b;

            if (c != 0)
                printf("a 0x%lx, b 0x%lx, c 0x%lx\n",a,b,c);
            else
                printf(".");
        }
        return (0);
}
```
```sh
$ zcc +test main.c -lfastmath; z88dk-ticks -x a.map a.bin
Decoded cdb: <F:G$main$0_0$0({0}DF,SI:S),C,0,0,0,0,0>
Decoded cdb: <S:Lmain.c.main$a$1_0$2({4}SL:U),B,1,-4>
Decoded cdb: <S:Lmain.c.main$b$1_0$2({4}SL:U),B,1,-8>
Decoded cdb: <S:Lmain.c.main$c$1_0$2({4}SL:U),B,1,-12>
Decoded cdb: <S:Lmain.c.main$i$1_0$2({1}SC:U),B,1,-13>
a 0x2, b 0x2, c 0x4
a 0x4, b 0x4, c 0x10
a 0x8, b 0x8, c 0x40
a 0x10, b 0x10, c 0x100
a 0x20, b 0x20, c 0x400
a 0x40, b 0x40, c 0x1000
a 0x80, b 0x80, c 0x4000
a 0x100, b 0x100, c 0x10000
a 0x200, b 0x200, c 0x40000
a 0x400, b 0x400, c 0x100000
a 0x800, b 0x800, c 0x400000
a 0x1000, b 0x1000, c 0x1000000
a 0x2000, b 0x2000, c 0x4000000
a 0x4000, b 0x4000, c 0x10000000
a 0x8000, b 0x8000, c 0x40000000
........a 0x1000000, b 0x1000000, c 0x1000100
a 0x2000000, b 0x2000000, c 0x2000200
a 0x4000000, b 0x4000000, c 0x4000400
a 0x8000000, b 0x8000000, c 0x8000800
a 0x10000000, b 0x10000000, c 0x10001000
a 0x20000000, b 0x20000000, c 0x20002000
a 0x40000000, b 0x40000000, c 0x40004000
a 0x80000000, b 0x80000000, c 0x80008000

Ticks: 1339982
```

**Comment by feilipu** [comment]
So, to fix it.

**Comment by suborb** [comment]
I can tell you're enjoying this one!

Any idea why newlib didn't exhibit the same issue?

**Comment by feilipu** [comment]
It might been having the same issue, just not tested?
It calls the same `l_mulu_32_32x32` linkage.

I'm pulling my hair out trying to unwrap the fast multiply stack tho...

**Comment by feilipu** [comment]
My strategy to resolve this is... as the multiply is done by adding component multiplies, each of the component multiplies must result in 0 (overflowed) or the end result would not be overflowed. So I've been disabling sections of the `_24b_24b` code inside `l_fast_mulu_32_32x32` to find which multiply is the culprit.

I've gotten as far as concluding that the `l_fast_mulu_32_24x16` is the problem multiply.
But, I haven't yet worked out where it is going wrong, or how it is being fed wrongness.

**Comment by feilipu** [comment]
Not happy I can't fix this.

Thinking it might be good to build a new solution, and see if I can hit same ticks count with a simpler code flow.
